### DISCUSSION TITLE
Client and server auth not synced

### DISCUSSION THREAD

**Discussion by jaredcosulich** [supabase/supabase#5795] [Questions]
Sometimes when I call: 

supabaseClient.auth.user() 

from the client I get a user back but when that user makes a request from the client to the server and I call: 

supabaseClient.auth.api.getUserByCookie(req, res) 

on the server I get an error saying "Invalid Refresh Token". 

How is this supposed to be handled? Am I missing documentation about this?

**Comment by egor-romanov** [comment]
Hi Jared,

Please take a look at the [docs](https://supabase.com/docs/reference/javascript/auth-user):

Are you using `auth.api.getUserByCookie()` according to it? I mean on the server-side to get the user? 
If I don't get you right, could you please share a snippet or repo? So I can understand how you are using it and what do you want to achieve.

**Reply by jaredcosulich** [reply]
Hi Egor, 

I am calling auth.api.getUserByCookie(req, res) on the server side immediately after calling supabaseClient.auth.user() on the client side (as mentioned in the original post).

I am not sure I understand why 

supabaseClient.auth.user()

is providing a user object but

auth.api.getUserByCookie(req, res)

is throwing an error.

An example of this might be:

```
useEffect(() => {
  const _user = supabaseClient.auth.user();
  if (_user) {
    [MAKE CALL TO SERVER]
  }
}, [user]
```

and then on the server function might look like:

```
export default async function handle(req, res) {
  const { user, error  } = await supabaseClient.auth.api.getUserByCookie(req, res)
  [PERFORM LOGIC INCLUDING ERROR HANDLING]
  res.json(result)
}
```

Despite there being a user returned on the clientside the server might throw an error with the message

"Invalid Refresh Token"

Is it recommended that I call signIn with a refresh token every time the user interacts with the website? If I do so then it seems to trigger an "onAuthStateChange" for the "SIGNED_IN" event even if the user is already logged in.

```
  const session = supabaseClient.auth.session()
  await supabaseClient.auth.signIn({
    refreshToken: session.refresh_token,
  });
```

This whole flow is quite awkward. I just want to know if the user is logged in or not on the clientSide so I can redirect them to a login page if necessary. Ideally I would be able to keep them logged in if they are actively using the website as well. I would think this would be basic functionality for an authentication flow. Am I missing something?

**Reply by egor-romanov** [reply]
Hey Jared,
Egor is here again

I'm sorry, didn't quite catch what was your problem exactly.
I think what you are looking for right now is this docs: https://supabase.com/docs/reference/javascript/auth-setauth

Hope it helps!

**Comment by egor-romanov** [comment]
May it be that you are looking for something like this:

https://supabase.com/docs/reference/javascript/auth-signin#sign-in-using-a-refresh-token-eg-in-react-native

**Reply by jaredcosulich** [reply]
This may be the solution, but it is quite awkward. How often should I be signing them in? When I do it triggers and `onAuthStateChange` even if they are already signed in which has to be handled properly. Is this the recommended workflow? 

Also this solution fails if the refreshToken has expired. Should I be looking for that failure to know if they are actually logged in or not?

**Reply by egor-romanov** [reply]
Hey again, I think I may have a clue on what you are looking for. Replied in another thread. [Here](https://github.com/supabase/supabase/discussions/5795#discussioncomment-2295169)

**Comment by jaredcosulich** [comment]
I'm sorry I'm not sure how this helps me know if the user is logged in or not when on the client side.

I also can not find an `X-Supabase-Auth` header in the request.

When I look at the headers in the request I see this:

```
console.log(req.rawHeaders)

output:
[
  'Host',
  'localhost:3000',
  'Connection',
  'keep-alive',
  'sec-ch-ua',
  '"(Not(A:Brand";v="8", "Chromium";v="100", "Google Chrome";v="100"',
  'sec-ch-ua-mobile',
  '?0',
  'User-Agent',
  'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/100.0.4896.20 Safari/537.36',
  'sec-ch-ua-platform',
  '"macOS"',
  'Content-Type',
  'application/json',
  'Accept',
  '*/*',
  'Sec-Fetch-Site',
  'same-origin',
  'Sec-Fetch-Mode',
  'cors',
  'Sec-Fetch-Dest',
  'empty',
  'Referer',
  'http://localhost:3000/',
  'Accept-Encoding',
  'gzip, deflate, br',
  'Accept-Language',
  'en-US,en;q=0.9',
  'Cookie',
  'sb-access-token=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE2NDYzOTg4NDYsInN1YiI6IjIzOGFhMTQxLTc2OWUtNGNhZC1iMjY4LTg0OTBlZjFhNzYxNiIsImVtYWlsIjoiamFyZWRAaXJyYXRpb25hbGRlc2lnbi5jb20iLCJwaG9uZSI6IiIsImFwcF9tZXRhZGF0YSI6eyJwcm92aWRlciI6ImVtYWlsIiwicHJvdmlkZXJzIjpbImVtYWlsIl19LCJ1c2VyX21ldGFkYXRhIjp7fSwicm9sZSI6ImF1dGhlbnRpY2F0ZWQifQ.NlEojKOoA__tiJ1Pq4ynpvmSm5tSwemAUOLcKo_QBWA; sb-refresh-token=WbdEv45lAxzvMkXAW2GiJA'
]
```

But maybe we can start with: what is the best way to know if the user is logged in or not on the client side?

**Reply by jaredcosulich** [reply]
Ok it looks like the documentation is suggesting that I add the `X-Supabase-Auth` header. Is the idea that every api call should include that header and call the `setAuth` function?

**Reply by egor-romanov** [reply]
Both `supabaseClient.auth.user()` and `supabaseClient.auth.session()` are valid to check if there is logged in user or active session.

**Comment by egor-romanov** [comment]
Yep. The idea is that you authorise as user on your client side. Then you can access this client's session like in the snippet below. And add access_token from it as header in request to your backend.

```
/**
* Make a request from the client to your server function
*/
async function makeApiRequest() {
  const token = newClient.session()?.access_token

  await fetch('https://example.com/withAuth', {
     method: 'GET',
     withCredentials: true,
     credentials: 'include',
     headers: {
      'Content-Type': 'application/json'
      'Authorization': bearer, // Your own auth
      'X-Supabase-Auth': token, // Set the Supabase user
     }
  })
}
```

On your backend side you will be able to auth with access_token, received from client.

```
/**
* Use the Auth token in your server-side function.
*/
async function apiFunction(req, res) {
  const { access_token } = req.get('X-Supabase-Auth') 

  // You can now use it within a Supabase Client 
  const supabase = createClient("https://xyzcompany.supabase.co", "public-anon-key")
  const { user, error } = supabase.auth.setAuth(access_token)

  // This client will now send requests as this user
  const { data } = await supabase.from('your_table').select()
}
```

That would be pretty restful as I see it. Because you won't store information on your backend between requests and just use token received in current request.

**Reply by jaredcosulich** [reply]
Ok, thank you for the help but this is still now working (or I'm missing something)

So right now on my client I get a user and a session when I call:

`supabaseClient.auth.user()` and `supabaseClient.auth.session()` respectively.

When I pass in the access_token that I have obtained from the session on the client side and call:

`const response = supabaseClient.auth.setAuth(access_token)`

on the server, though, I get:

```
{
  access_token: [ACCESS TOKEN],
  token_type: 'bearer',
  user: null
}
```

So I am still getting a user and a session on the client that does not allow me to interact with the server...

**Reply by egor-romanov** [reply]
Could you create a simple example and publish it on GitHub so I can reproduce this behaviour?
You may use [this example as the template](https://github.com/supabase/supabase/tree/master/examples/nextjs-with-supabase-auth).

In the mean time we have [this issue](https://github.com/supabase/supabase/issues/1735), where you can find some examples for solving similar problem.

**Reply by jaredcosulich** [reply]
Ok I will try to find time to put together a complete example.

It sounds like this should not happen. Is that correct? 

If I take the access_token provided by `supabaseClient.auth.session()` and pass that to the server it should allow me to interact with Supabase from the server as if I were that user by calling `supabaseClient.auth.setAuth(access_token)`?

Thank you for the help. I will read through the issue link you've provided.

**Reply by egor-romanov** [reply]
> It sounds like this should not happen. Is that correct?

> If I take the access_token provided by supabaseClient.auth.session() and pass that to the server it should allow me to interact with Supabase from the server as if I were that user by calling supabaseClient.auth.setAuth(access_token)?

Yes, that's true

**Reply by jaredcosulich** [reply]
Ok, thank you for clarifying that.
